### DISCUSSION TITLE
Email update flow

### DISCUSSION THREAD

**Discussion by woylie** [ory/kratos#2501] [Q&A]
I've set up email/password credentials and email verification. At the moment, when a user updates the email address, it is changed immediately. If there is a typo in the new address, the user won't be able to log in with the new address (because it is not verified) or the old address (because the value has already changed). I would like to change the flow like this:

1. user submits updated email address
2. the identifier remains the old email address until the new one is verified
3. user verifies new email address
4. identifier is changed to new email address

An alternative flow might be to use an array field for emails:

1. user adds a new email address
2. user verifies new address
3. user marks new address as primary
4. user optionally removes old email address
5. user can only remove email addresses if there is at least one other verified address (or when it is not the primary address)

Is there any way to implement one of the flows currently? For the first flow, it is unclear to me how you would _not_ change the address until it is verified. For the second flow, it is unclear to me a) how you would enforce that there is at least one verified address during the settings flow, and b) how you would mark one (and only one) of the addresses as primary, which would be important when sending out email notifications.

**Comment by aeneasr** [comment]
Thank you! That is indeed a good point. We had some ideas about this, the flow is quite difficult to implement from a UX perspective. That is why this is currently on the backlog and not worked on actively.

By the way it’s also a security problem! Because you can use this right now to perform account enumeration.

Probably relevant: https://github.com/ory/kratos/issues/140

**Reply by woylie** [reply]
Is there anything you can share from previous considerations? I find this fairly important and would like to put some more thought into this later this week.

**Comment by woylie** [comment]
## Single email field

### General flow

1. user submits updated email address
2. the identifier remains the old email address until the new one is verified
3. user verifies new email address
4. identifier is changed to new email address

### What can be done right now

Until Kratos supports this flow, one could think about handling the whole change
and verification flow outside of Kratos and using its API to finally update the
identifier. However, there is currently no way of updating the verification
status via the API, so this seems impossible.

### What needs to be changed in Kratos

#### Workaround: handling flow outside of Kratos

If the Kratos API is changed to allow setting the verification status of a
verifiable address (#2473), the whole identifier change flow could be handled
in a server application.

1. hide identifier fields in settings UI based on field name
2. use a blocking Webhook (#1585) to prevent the email/identifier from being
   changed during the settings flow (requires that an identity can be patched,
   so that the email can be removed from the payload, or that the previous
   identity traits are passed as parameters to the webhook, so that the email
   field can be overridden with the previous value)
3. implement custom identifier-change form and backend logic for saving
   generating confirmation link, storing flow data in external DB, sending
   confirmation link and confirmation endpoint
4. on confirmation, update the email field via Kratos API and set `verified` to
   `true`

#### Solution: add verifiable_address method to settings flow

1. ensure `profile` form fields don't include any verifiable traits, and a
   submit with method `profile` does not update any verifiable traits
2. render form fields for verifiable traits as separate form groups (one per
   verifiable trait)
   - trait (hidden input) - the field name, e.g. `email`
   - value (text input)
   - submit button (name: `method`, value: `verifiable_address`)
3. for all verifiable traits with pending changes
   - current value (disabled text input or text node)
   - pending value (disabled text input or text node)
   - text node informing the user that the verification for the new address is
     pending
   - potentially buttons to cancel the update/verification or to re-send
     the verification link

From just looking at the DB tables, I'm not sure how the verification flow
works regarding where the address is stored that is being verified. It seems
like in the `selfservice_verification_flow` table, the address is only stored
in the `ui` column. We need a way to store the new value without updating the
old value until the new address is verified. I would imagine something like
this:

1. add a `replaces` FK column to the `identity_verifiable_address` table
2. when the settings flow is submitted with the `verifiable_address` method,
   initiate verification flow
   - store submitted value as new row in the `identity_verifiable_address` table
   - set `replaces` to the ID of the previous address
3. when a verification flow is completed and the `replaces` field is set
   1. set `verified` to `true` for that address as usual
   2. update identity trait to the new value
   3. remove previous verifiable address, set `replaces` to `NULL` on new
      address

## Primary email field + secondary email(s) field

### What can be done right now

1. add one `primary_email` field (string) and one `secondary_email` (string)
   or `additional_emails` (array) field to the identity traits, both with
   enabled verification
2. disable or hide input for primary email in UI
3. the secondary email addresses can be handled normally
4. use blocking webhook (#1585) to prevent direct updates to the primary
   email field during the settings flow
5. build custom UI and custom backend logic outside of Kratos to allow a user
   to turn the secondary email or one of the additional emails into the
   primary email (ensuring only verified addresses can be selected)
6. update the traits using the Kratos API, moving the previous primary address
   to the secondary emails and the new primary address from secondary emails
   to primary email

### What can be changed to make this easier

#### Native support

Native support for this flow would certainly be nice.

On the configuration side, this might be enabled by setting a `secondary` value on the `verification` options pointing to the trait with the secondary values:

```json
{
  "properties": {
    "traits": {
      "type": "object",
      "properties": {
        "primary_email": {
          "type": "string",
          "format": "email",
          "ory.sh/kratos": {
            "verification": {
              "via": "email",
              "secondary": "secondary_emails"
            }
          }
        },
        "secondary_emails": {
          "type": "array",
          "items": {
            "type": "string"
          }
        },
      }
    }
  }
}
```

I'm not sure how you would return the UI nodes for the settings flow, though. We would need to list each secondary email, probably as text node, and add a button to each verified one to promote it to primary. And there probably one additional empty input for adding a new email/identifier. For this flow, it would be sufficient to only support adding and removing secondary addresses, and selecting a verified secondary address as primary. We don't need to be able to update existing addresses.

#### Immutable verifiable addresses

Even if this flow is not supported natively, the workaround described above doesn't seem so bad, except for the blocking webhook. It would be nicer if an identifier field could be marked as immutable during the settings flow. We have `metadata_public` now, but I don't think we can add an identifier there at the moment, since then it wouldn't appear in the registration flow. Since this only relates to verification, and not identifiers per se, I guess an option could be to add another verification option.

```json
{
  "properties": {
    "traits": {
      "type": "object",
      "properties": {
        "primary_email": {
          "type": "string",
          "format": "email",
          "ory.sh/kratos": {
            "verification": {
              "via": "email",
              "immutable": true
            }
          }
        }
      }
    }
  }
}
```

The expected behavior would be to show a form field in the registration flow, but not in the settings flow.

---

What do you think? @aeneasr

**Reply by nerrixde** [reply]
any updates on this? @aeneasr?

**Reply by aeneasr** [reply]
No. The feature does not have a high priority for Ory right now. If there’s progress it will probably be here: https://github.com/ory/kratos/issues/140
